### PR TITLE
Typography migration

### PR DESCRIPTION
**Pull request by mihajlovco**
With this PR we will migrate the legacy theme typography styles to a new structure and style expressions are migrated to the new type of access.

## Features
- Find legacy typography objects in Apps/Theme/theme.ts and map them to new typography styles.
- Migrate all theme expressions for `typography`, `colors`, `fonts`, `breakpoints` and `borderRadius` to match the new theme styles access
- Remove theme imports
- User will receive a message if typography declaration is not found.
- Support for custom typography styles migration.
- User will receive information message if the typography styles are `fully` or `partially migrated`. Styles names of the `migrated typography styles will be listed` and `migration canceled`.  That will get the attention of the developers in the case when they want to `upgrade an already upgraded project after the merge that does not contain the backup folder` for detection.
- The user will receive information if a `custom style structure is detected` in the typography styles. That means instead of the Webiny standard CSS object like `heading: { fontColor: 10 }` users have `Array`, `Map`, or other iterable types. In this case, we can't map the typography styles.
- User will receive information for creating a backup folder in `apps/_theme_backup`. If it exists will cancel the migration.
- The user will receive a message if some of the custom style keys can't be found in any typography type like `headings`, `paragraphs`, `lists`, and `quotes` during style expression migration like `${theme.styles.typography.myTypoErrorPropKey}` which will lead to `typo errors detection` of the style keys. 
In this case, the `expression will not be migrated` and a warning message will display the `style key name` and `line number in the file` so the developer can navigate quickly and manually write the expression.

- The user will receive a message that the custom name of the typography property can't be detected for the type of the `HTML tag`. So by default will be set to `p` or `paragraph`. The user will be asked to manually reorganize the files.

## Migrated typography object sample
```
export const typography = {
    paragraphs: [ { id: "paragraph1", name: "paragraph1", tag: "p", css: { ...paragraphs, fontSize: 16.5 } } ],
    lists: [{ id: "list", name: "list", tag: "ul", css: { ...paragraphs, fontSize: 17 } }],
};
```

### Migrated pages
Location: `apps/theme/layouts/pages/Static`:
- Footer
- Header
- HeaderDesktop
- HeaderMobile
- Navigation

#### Expressions
- typography -> `props => props.theme.styles.typography.${typographyType}.cssById("${styleKey}")`
- colors -> `props => props.theme.styles.colors["${colorName}"]`
- fonts -> `props => props.theme.fonts["${fontName}"]`
- breakpoints -> `props => props.theme.breakpoints["${breakPointAccessName}"]`
- single literal string expression like `borderRadius` or  `breakpoints` -> `props => props.theme.${literalName}`

#### Imports
* Removed `theme` import declaration line on all pages. 

### Backup
Backup for all files in the `apps/theme`  is created in the new folder  `apps/_theme_backup`. All files will be copied before any migration is started.

> Note: This backup will contain all the files for `apps/theme/layouts/forms` that need to cover `updateDefaultFormLayout` migration process. The creation of the backup folder in `updateDefaultFormLayout` migration process is removed. 

## Process
theme-typography-migration

## Upgrade testing video - tested on our repo 
https://user-images.githubusercontent.com/82515066/232148295-992f67ff-c016-466c-8e1c-6858a6470802.mov

## Test with installed new webiny project version 5.34
test-webiny-deployed-v5 34

### CODE REVIEW

File: utils/classes/FileDefinition.ts [hunk 1]
```diff
@@ -1,4 +1,16 @@
...
+    | "cms"
+    | "gql"
+    | "ps"
+    | "ddb2es"
+    | "theme"
+    | "website"
+    | "admin"
+    | "core"
+    | "fb"
+    | "cwp";
```
Comment (adrians5j) [inline utils/classes/FileDefinition.ts]: Is `cwp` really needed? I don't think so, but maybe I'm missing smt. 🤔

File: utils/createMorphProject.ts [hunk 2]
```diff
@@ -3,7 +3,11 @@ import { Project } from "ts-morph";
 export const createMorphProject = (files: string[]): Project => {
     const project = new Project();
     for (const file of files) {
-        project.addSourceFileAtPath(file);
+        try {
+            project.addSourceFileAtPath(file);
+        } catch (e) {
+            console.log("file can't be found. File path: ", file);
```
Comment (adrians5j) [inline utils/createMorphProject.ts]: Let's use log from `import log from "./log";` here. Then it's:

log.warning("Could not add %s file.", file);

File: 5.35.0/index.ts [hunk 3]
```diff
@@ -2,28 +2,28 @@ import { createMorphProject, prettierFormat, yarnInstall } from "../utils";
...
     const files = setupFiles(context);
     const rawFiles = files.paths();
     const project = createMorphProject(rawFiles);
-
-    const processors = [updateGraphQL, updateAdminApp, updateToEmotion11, updateDefaultFormLayout];
-
-    for (let i = 0; i < processors.length; i++) {
-        const processor = processors[i];
-        await processor({
+    await [
```
Comment (adrians5j) [inline 5.35.0/index.ts]: Hm, why we have `reduce`-version of the code here back?

Is there any reason we undo-ed what was done here in one of my previous commits?

File: 5.35.0/themeTypographyMigration/createThemeUpgradeBackup.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            fs.mkdirSync(toBackup);
+            await ncp(from, toBackup);
+            results[0] = true;
+
+            const packageJsonPath = path.join(to, "package.json");
+            const packageJson = fs.readFileSync(packageJsonPath, "utf8");
+
+            fs.writeFileSync(
+                packageJsonPath,
+                packageJson.replace(/"\^latest"/g, `"${currentProjectVersion}"`)
```
Comment (adrians5j) [inline 5.35.0/themeTypographyMigration/createThemeUpgradeBackup.ts]: Hm, unless I'm missing smt, this step should not be needed.

File: 5.35.0/migrateThemeTypography.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    const { context, project } = params;
+
+    context.log.info(`Upgrading Webiny theme to new typography styles...`);
+
+    context.log.debug(`Checking the legacy theme object...`);
+    // Get the theme file and typography object
+    const themeSourceFile = getAppThemeSourceFile(context, project);
+
+    if (!themeSourceFile) {
+        context.log.info(
```
Comment (adrians5j) [inline 5.35.0/migrateThemeTypography.ts]:
```suggestion
        context.log.warning(
```

File: 5.35.0/migrateThemeTypography.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+            `Theme typography styles upgrade is canceled. Can't find App/Theme/theme.ts file`
+        );
+        return;
+    }
+
+    const typographyVariable = getTypographyVariableDeclaration(themeSourceFile);
+    if (!typographyVariable) {
+        context.log.info(
+            `Theme's typography styles can't be found in App/Theme/theme.ts file. Do you have custom implementation?
```
Comment (adrians5j) [inline 5.35.0/migrateThemeTypography.ts]: Double check this message.

File: 5.35.0/migrateThemeTypography.ts [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+     * MIGRATE THE THEME OBJECT
+     */
+    context.log.debug(`Map legacy typography object to new structure...`);
+    const typographyMappingResult = mapToNewTypographyStyles(typographyVariable, context);
+
+    if (!typographyMappingResult.isSuccessfullyMapped) {
+        context.log.info(typographyMappingResult.info);
+        return;
+    }
+    context.log.debug(`Legacy typography object is successfully mapped.`);
```
Comment (adrians5j) [inline 5.35.0/migrateThemeTypography.ts]:
```suggestion
    context.log.success(`Legacy typography object is successfully mapped.`);
```

File: 5.35.0/migrateThemeTypography.ts [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    // for user's custom style keys
+
+    const styleIdToTypographyTypeMap = typographyMappingResult.styleIdTopographyType;
+    context.log.debug(`Migrate solution files...`);
+    const migrationDefinitions = themeMigrationFilesSetup(context);
+    for (const definition of migrationDefinitions) {
+        migrateFile(definition, styleIdToTypographyTypeMap, project, context);
+    }
+
+    context.log.success(`Cheers! You have successfully upgraded to new Webiny typography styles.`);
```
Comment (adrians5j) [inline 5.35.0/migrateThemeTypography.ts]: Let's just be more in tone with other messages that we have here. There's gonna be a lot of these logs, so let's make them brief and informative.

In this case, 

```
context.log.success("Typography styles migrated.");
```

(or something along those lines)

**Review by adrians5j — CHANGES_REQUESTED** [review]
Apart from the comments below, please check the following:

1. Issue When Running the Upgrade ⬇️ 
image

2. Order of Execution Incorrect
Please check the `5.35.0/index.ts` comments. This is why the above logs are coming out in an incorrect order (instead of steps being execute one after the other, steps are all executed in parallel, which should not happen).

3. Your Code Should Execute Before "updateDefaultFormLayout"
In your code, you create a backup of the theme, but that's happening after the `updateDefaultFormLayout` step.

Because of this, the backup that gets created contains the new form code, not user's existing one. Let's ensure the backing up of `theme` folder happens before any other code changes are applied.


Thanks 🍻

File: 5.35.0/themeTypographyMigration/createThemeUpgradeBackup.ts (hunk 4 above)
Comment (mihajlovco) [reply 5.35.0/themeTypographyMigration/createThemeUpgradeBackup.ts]: This is the code that creates the backup for the `theme` in folder "`_theme_backup` " , let's sync tomorrow to resolve this issue.

File: 5.35.0/index.ts (hunk 3 above)
Comment (mihajlovco) [reply 5.35.0/index.ts]: I have merged the code from the main branch and for sure I did a mistake, I didn't undo anything deliberately. Let's sync tomorrow for this issue as well.

File: utils/classes/FileDefinition.ts (hunk 1 above)
Comment (mihajlovco) [reply utils/classes/FileDefinition.ts]: I set the old tag definitions, I don't need other tag names anymore I am using only the `theme` now. 
Thanks.

File: 5.35.0/index.ts (hunk 3 above)
Comment (mihajlovco) [reply 5.35.0/index.ts]: I resolved this issue, I set the new code, works well now upgrades are executing in sequence.

**Pull request merged**